Link error due to missing -lstdc++fs (e.g. on CentOS 8.2)

Same error with CentOS 8.2.

Trying to configure with:
./configure --with-crypto-library=sodium --with-bundled-catch --with-bundled-pegtl

Getting error:
....
configure: Using bundled Catch library
checking tao/pegtl.hpp usability... no
checking tao/pegtl.hpp presence... no
checking for tao/pegtl.hpp... no
configure: error: in `/home/username/rpmbuild/BUILD/usbguard-0.7.8':
configure: error: PEGTL header file not found or not usable. Re-run with --with-bundled-pegtl to use the bundled library.

Also tried compiling PEGTL without any success. Official USBGuard release in CentOS8 is not stable. Really need to this.

Please help or point me in any direction!

Best regards

The `PEGTL header file not found or not usable` goes away once you install package `gcc-c++`. But there will be link errors in the end:

```console
# make V=1    # link error in isolation
Makefile:3498: warning: overriding recipe for target 'check'
Makefile:3004: warning: ignoring old recipe for target 'check'
make  all-recursive
Makefile:3498: warning: overriding recipe for target 'check'
Makefile:3004: warning: ignoring old recipe for target 'check'
make[1]: Entering directory '/usbguard'
Making all in src/Tests/
make[2]: Entering directory '/usbguard/src/Tests'
Making all in Fuzzers
make[3]: Entering directory '/usbguard/src/Tests/Fuzzers'
make[3]: Nothing to be done for 'all'.
make[3]: Leaving directory '/usbguard/src/Tests/Fuzzers'
make[3]: Entering directory '/usbguard/src/Tests'
make[3]: Nothing to be done for 'all-am'.
make[3]: Leaving directory '/usbguard/src/Tests'
make[2]: Leaving directory '/usbguard/src/Tests'
Makefile:3498: warning: overriding recipe for target 'check'
Makefile:3004: warning: ignoring old recipe for target 'check'
make[2]: Entering directory '/usbguard'
/bin/sh ./libtool  --tag=CXX   --mode=link g++ -pthread -std=c++17   -fvisibility=hidden  -pedantic -Wno-unknown-pragmas -Wall -Wextra -Wformat=2 -Wredundant-decls -Wcast-align -Wmissing-declarations -Wmissing-include-dirs -Wmissing-format-attribute -Wswitch-enum -Wswitch-default -Winvalid-pch -Wformat-nonliteral -Wno-deprecated-register -Wno-missing-field-initializers -Wno-implicit-fallthrough    -o usbguard src/CLI/usbguard-usbguard.o src/CLI/usbguard-usbguard-list-devices.o src/CLI/usbguard-usbguard-allow-device.o src/CLI/usbguard-usbguard-block-device.o src/CLI/usbguard-usbguard-reject-device.o src/CLI/usbguard-usbguard-apply-device-policy.o src/CLI/usbguard-usbguard-list-rules.o src/CLI/usbguard-usbguard-append-rule.o src/CLI/usbguard-usbguard-remove-rule.o src/CLI/usbguard-usbguard-generate-policy.o src/CLI/usbguard-usbguard-watch.o src/CLI/usbguard-IPCSignalWatcher.o src/CLI/usbguard-PolicyGenerator.o src/CLI/usbguard-usbguard-read-descriptor.o src/CLI/usbguard-usbguard-set-parameter.o src/CLI/usbguard-usbguard-get-parameter.o src/CLI/usbguard-usbguard-add-user.o src/CLI/usbguard-usbguard-remove-user.o ./libusbguard.la  
libtool: link: g++ -pthread -std=c++17 -fvisibility=hidden -pedantic -Wno-unknown-pragmas -Wall -Wextra -Wformat=2 -Wredundant-decls -Wcast-align -Wmissing-declarations -Wmissing-include-dirs -Wmissing-format-attribute -Wswitch-enum -Wswitch-default -Winvalid-pch -Wformat-nonliteral -Wno-deprecated-register -Wno-missing-field-initializers -Wno-implicit-fallthrough -o .libs/usbguard src/CLI/usbguard-usbguard.o src/CLI/usbguard-usbguard-list-devices.o src/CLI/usbguard-usbguard-allow-device.o src/CLI/usbguard-usbguard-block-device.o src/CLI/usbguard-usbguard-reject-device.o src/CLI/usbguard-usbguard-apply-device-policy.o src/CLI/usbguard-usbguard-list-rules.o src/CLI/usbguard-usbguard-append-rule.o src/CLI/usbguard-usbguard-remove-rule.o src/CLI/usbguard-usbguard-generate-policy.o src/CLI/usbguard-usbguard-watch.o src/CLI/usbguard-IPCSignalWatcher.o src/CLI/usbguard-PolicyGenerator.o src/CLI/usbguard-usbguard-read-descriptor.o src/CLI/usbguard-usbguard-set-parameter.o src/CLI/usbguard-usbguard-get-parameter.o src/CLI/usbguard-usbguard-add-user.o src/CLI/usbguard-usbguard-remove-user.o  ./.libs/libusbguard.so -lqb -lprotobuf -lpthread -lsodium -pthread -Wl,-rpath -Wl,/usr/local/lib
./.libs/libusbguard.so: undefined reference to `vtable for std::filesystem::__cxx11::filesystem_error'
./.libs/libusbguard.so: undefined reference to `typeinfo for std::filesystem::__cxx11::filesystem_error'
./.libs/libusbguard.so: undefined reference to `std::filesystem::__cxx11::filesystem_error::~filesystem_error()'
./.libs/libusbguard.so: undefined reference to `std::filesystem::__cxx11::path::_M_split_cmpts()'
./.libs/libusbguard.so: undefined reference to `std::filesystem::__cxx11::filesystem_error::_M_gen_what()'
collect2: error: ld returned 1 exit status
make[2]: *** [Makefile:1450: usbguard] Error 1
make[2]: Leaving directory '/usbguard'
make[1]: *** [Makefile:2712: all-recursive] Error 1
make[1]: Leaving directory '/usbguard'
make: *** [Makefile:1044: all] Error 2
The command '/bin/sh -c make V=1    # link error in isolation' returned a non-zero code: 2
```

Here's my Dockerfile to see that error in action:
```Dockerfile
# Copyright (c) 2022 Sebastian Pipping <sebastian@pipping.org>
# Licensed under the Apache license version 2.0

FROM centos:8.2.2004
RUN dnf install -y epel-release dnf-plugins-core \
        && \
    dnf config-manager --set-enabled PowerTools \
        && \
    dnf install -y \
        autoconf \
        automake \
        dbus-glib-devel \
        diffutils \
        file \
        gcc-c++ \
        git \
        libqb-devel \
        libgcrypt-devel \
        libsodium-devel \
        libtool \
        libxslt \
        make \
        polkit-devel \
        protobuf-devel \
        protobuf-compiler
RUN git clone --depth 1 https://github.com/USBGuard/usbguard
WORKDIR usbguard
RUN ./autogen.sh
RUN ./configure --with-crypto-library=sodium --with-bundled-catch --with-bundled-pegtl || ! cat config.log
RUN ! make V=1 "-j$(nproc)"  # until the link error
RUN make V=1  # link error in isolation
```

--- a/.github/workflows/docker.yml
+++ b/.github/workflows/docker.yml
@@ -1,0 +1,48 @@
+##
+## Copyright (c) 2022 Sebastian Pipping <sebastian@pipping.org>
+##
+## This program is free software; you can redistribute it and/or modify
+## it under the terms of the GNU General Public License as published by
+## the Free Software Foundation; either version 2 of the License, or
+## (at your option) any later version.
+##
+## This program is distributed in the hope that it will be useful,
+## but WITHOUT ANY WARRANTY; without even the implied warranty of
+## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+## GNU General Public License for more details.
+##
+## You should have received a copy of the GNU General Public License
+## along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+name: Build on non-Ubuntu Linux using Docker
+
+on:
+  push:
+  pull_request:
+
+jobs:
+  build_docker:
+    strategy:
+      fail-fast: false
+      matrix:
+        linux_distro:
+        - CentOS 8.2         # with GCC 8.5.0
+
+    name: Build on ${{ matrix.linux_distro }} using Docker
+    runs-on: ubuntu-20.04
+
+    steps:
+    - uses: actions/checkout@v2
+
+    - name: Build on ${{ matrix.linux_distro }} using Docker
+      run:  |
+        set -x
+
+        # Create tarballs used by Dockerfile
+        git submodule update --init
+        git archive -o usbguard.tar HEAD
+        ( cd src/ThirdParty/PEGTL/ && git archive -o ../../../pegtl.tar HEAD )
+
+        # Build using Docker
+        linux_distro="$(tr '[:upper:]' '[:lower:]' <<<"${{ matrix.linux_distro }}" | sed 's,[ .],_,g')"
+        docker build -f scripts/docker/build_on_"${linux_distro}".Dockerfile .

--- a/Makefile.am
+++ b/Makefile.am
@@ -28,6 +28,7 @@ EXTRA_DIST =\
 	VERSION \
 	CHANGELOG.md \
 	src/astylerc \
+	src/test_filesystem.cpp \
 	scripts/usbguard-zsh-completion \
 	scripts/modeline.vim \
 	scripts/astyle.sh \

--- a/configure.ac
+++ b/configure.ac
@@ -362,19 +362,43 @@ AC_SUBST([catch_CFLAGS])
 AC_SUBST([catch_LIBS])
 
 #
+# stdc++fs (for PEGTL >=3 below, e.g. with GCC 8)
+#
+STDCXX_FS_LINKTEST_FILENAME="${srcdir}"/src/test_filesystem.cpp
+STDCXX_FS_LINKTEST_SOURCE="$(cat "${STDCXX_FS_LINKTEST_FILENAME}")"
+AC_LANG_PUSH([C++])
+AC_MSG_CHECKING([whether we need to link to -lstdc++fs for PEGTL explicitly])
+AC_LINK_IFELSE([AC_LANG_SOURCE([${STDCXX_FS_LINKTEST_SOURCE}])], [
+    AC_MSG_RESULT([no])
+    stdcxxfs_LIBS=''
+], [
+    stdcxxfs_LIBS='-lstdc++fs'
+    SAVE_LIBS=${LIBS}
+    LIBS="${LIBS} ${stdcxxfs_LIBS}"
+    AC_LINK_IFELSE([AC_LANG_SOURCE([${STDCXX_FS_LINKTEST_SOURCE}])], [
+        AC_MSG_RESULT([yes])
+    ], [
+        AC_MSG_RESULT([ERROR])
+        AC_MSG_ERROR([Link test failed both with and without ${stdcxxfs_LIBS}; something is broken, please check file config.log for details.])
+    ])
+    LIBS=${SAVE_LIBS}
+])
+AC_LANG_POP()
+
+#
 # PEGTL C++ library
 #
 AC_ARG_WITH([bundled-pegtl], AS_HELP_STRING([--with-bundled-pegtl], [Build using the bundled PEGTL library]), [with_bundled_pegtl=$withval], [with_bundled_pegtl=no])
 if test "x$with_bundled_pegtl" = xyes; then
 	pegtl_CFLAGS="-I\$(top_srcdir)/src/ThirdParty/PEGTL/include"
 	pegtl_AC_CFLAGS="-I$srcdir/src/ThirdParty/PEGTL/include"
-	pegtl_LIBS=""
+	pegtl_LIBS="${stdcxxfs_LIBS}"
 	AC_MSG_NOTICE([Using bundled PEGTL library])
 	pegtl_summary="bundled; $pegtl_CFLAGS $pegtl_LIBS"
 else
 	pegtl_CFLAGS=""
 	pegtl_AC_CFLAGS=""
-	pegtl_LIBS=""
+	pegtl_LIBS="${stdcxxfs_LIBS}"
 	pegtl_summary="system-wide; $pegtl_CFLAGS $pegtl_LIBS"
 fi
 AC_SUBST([pegtl_CFLAGS])

--- a/scripts/docker/build_on_centos_8_2.Dockerfile
+++ b/scripts/docker/build_on_centos_8_2.Dockerfile
@@ -1,0 +1,51 @@
+##
+## Copyright (c) 2022 Sebastian Pipping <sebastian@pipping.org>
+##
+## This program is free software; you can redistribute it and/or modify
+## it under the terms of the GNU General Public License as published by
+## the Free Software Foundation; either version 2 of the License, or
+## (at your option) any later version.
+##
+## This program is distributed in the hope that it will be useful,
+## but WITHOUT ANY WARRANTY; without even the implied warranty of
+## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+## GNU General Public License for more details.
+##
+## You should have received a copy of the GNU General Public License
+## along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+FROM centos:8.2.2004
+RUN sed \
+            -e '/^baseurl=.*/d' \
+            -e 's,^mirrorlist=.*repo=\([^&]\+\).*,baseurl=http://linuxsoft.cern.ch/centos-vault/8.2.2004/\1/x86_64/os/,' \
+            -i /etc/yum.repos.d/CentOS-* \
+        && \
+    dnf clean metadata \
+        && \
+    dnf install -y epel-release dnf-plugins-core \
+        && \
+    dnf config-manager --set-enabled PowerTools \
+        && \
+    dnf install -y \
+            autoconf \
+            automake \
+            dbus-glib-devel \
+            diffutils \
+            file \
+            gcc-c++ \
+            git \
+            libgcrypt-devel \
+            libqb-devel \
+            libsodium-devel \
+            libtool \
+            libxslt \
+            make \
+            polkit-devel \
+            protobuf-compiler \
+            protobuf-devel
+ADD usbguard.tar usbguard/
+ADD pegtl.tar usbguard/src/ThirdParty/PEGTL/
+WORKDIR usbguard
+RUN git init &>/dev/null && ./autogen.sh
+RUN ./configure --with-bundled-catch --with-bundled-pegtl || ! cat config.log
+RUN make V=1 "-j$(nproc)"

--- a/src/Tests/Source/CheckScripts/private-with-build-config.sh
+++ b/src/Tests/Source/CheckScripts/private-with-build-config.sh
@@ -6,6 +6,7 @@
 # check-path-include: *.h
 # check-path-exclude: src/Library/public/*.hpp
 # check-path-exclude: src/Tests/*
+# check-path-exclude: src/test_filesystem.cpp
 #
 SOURCE_FILEPATH="$1"
 

--- a/src/test_filesystem.cpp
+++ b/src/test_filesystem.cpp
@@ -1,0 +1,24 @@
+//
+// Copyright (c) 2022 Sebastian Pipping <sebastian@pipping.org>
+//
+// This program is free software; you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation; either version 2 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+#include <filesystem>
+
+int main()
+{
+  return std::filesystem::temp_directory_path().is_absolute();
+}
+
+/* vim: set ts=2 sw=2 et */

